server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java

```diff
@@ -119,61 +134,120 @@ public void exit() {
   enum Status {
     UNINITIALIZED,
     INITIALIZED,
-    SHUTDOWN;
+    SHUTDOWN
   }
 
   /**
-   * Try to update the Gradle Java home if:
-   * <ul>
-   *   <li>Gradle Java home is not set.</li>
-   *   <li>A valid JDK can be found to launch Gradle.</li>
-   * </ul>
-   *
-   * <p>The JDK installation path string will be set to {@link Preferences#gradleJavaHome}.
+   * Finds a suitable version of JDK to use for gradle operations.
+   * Returns {@code null} if no suitable JDK can be found.
    */
-  private void updateGradleJavaHomeIfNecessary(URI rootUri) {
+  private File getSuitableJdk(URI rootUri) {
+
     Preferences preferences = preferenceManager.getPreferences();
-    if (preferences.getJdks() == null || preferences.getJdks().isEmpty()) {
-      return;
+    String filePath = preferences.getGradleJavaHome();
+    File gradleJavaHome;
+    if (StringUtils.isNotEmpty(filePath)) {
+      gradleJavaHome = new File(filePath);
+    } else {
+      gradleJavaHome = connector.getGradleJavaHome(rootUri);
+    }
+
+    GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
+    Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
+    LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
+
+    String gradleVersion;
+    if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
+      gradleVersion = preferences.getGradleVersion();
+    } else {
+      gradleVersion = connector.getGradleVersion(rootUri);
     }
 
-    if (StringUtils.isBlank(preferences.getGradleJavaHome())) {
-      String gradleVersion = "";
-      GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
-      Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
-      LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
-      if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
-        gradleVersion = preferences.getGradleVersion();
-      } else {
-        gradleVersion = connector.getGradleVersion(rootUri);
+    if (StringUtils.isEmpty(gradleVersion)) {
+      return null;
+    }
+
+    map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
+    LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
+
+    String latestCompatibleVersion = Utils.getLatestCompatibleJavaVersion(gradleVersion);
+    String oldestCompatibleVersion = Utils.getOldestCompatibleJavaVersion();
+
+    if (StringUtils.isNotBlank(latestCompatibleVersion)) {
+
+      // Use GradleJavaHome if compatible
+      if (gradleJavaHome != null) {
+
+        try {
+          String gradleJavaHomeVersion = JavaUtils.getJavaVersionFromFile(gradleJavaHome);
```

Comment:
What will happen if, `JavaUtils.getJavaVersionFromFile` returns a string which is not a purely string? Will it throw exception in the `JavaUtils.isCompatible()`? Will the exception be caught?

Reply:
What do you mean by purely string? `JavaUtils#getJavaVersionFromFile` returns a string containing the version of the jdk or `null` upon failure. I will add a null check to prevent `NullPointerException` in `JavaUtils#isCompatible`.

Reply:
Sorry, my typo. 

I mean, for example, assume `JavaUtils.getJavaVersionFromFile` returns "foo17.0.1"

Reply:
Is that a special case in which the version string contains additional characters?
For example if we run the command `java -version`, the supposed output would be in the given format correct?
```
java -version
java version "21.0.3" 2024-04-16 LTS
```
And now we are extracting the version printed between the double quotes.

Reply:
I'm not sure. Since you know there are so many different java distribution, I'm not sure if they will all print the version like that.

I mean. no need to cover all those cases. If we found that the version string cannot be correctly parsed, just makes sure the server will not fail down due to the unhandled exceptions.

Reply:
Addressed in commit [e7de6d8](https://github.com/microsoft/build-server-for-gradle/pull/165/commits/e7de6d870c1893f46e4d25f1bfb2b59ffed35218)